scripts/APIDiffReport/Sources/APIDiffReport/diffreport.swift

```diff
@@ -0,0 +1,342 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ 
+ Code modified by Mapbox to adjust API filetring criteria, original at https://github.com/material-motion/apidiff/blob/e78f92ae310cd4affc86a4510bb7b9f9609662d2/apple/diffreport/Sources/diffreportlib/diffreport.swift
+ 
+ TODO: log briefly all code updates
+ */
+
+import Foundation
+
+public typealias JSONObject = Any
+
+typealias SourceKittenNode = [String: Any]
+typealias APINode = [String: Any]
+typealias ApiNameNodeMap = [String: APINode]
+
+/** A type of API change. */
+public enum ApiChange {
+    case addition(apiType: String, name: String)
+    case deletion(apiType: String, name: String)
+    case modification(apiType: String, name: String, modificationType: String, from: String, to: String)
+}
+
+/** Generates an API diff report from two SourceKitten JSON outputs. */
+public func diffreport(oldApi: JSONObject, newApi: JSONObject) throws -> [String: [ApiChange]] {
+    let oldApiNameNodeMap = extractAPINodeMap(from: oldApi as! [SourceKittenNode])
+    let newApiNameNodeMap = extractAPINodeMap(from: newApi as! [SourceKittenNode])
+    
+    let oldApiNames = Set(oldApiNameNodeMap.keys)
+    let newApiNames = Set(newApiNameNodeMap.keys)
+    
+    let addedApiNames = newApiNames.subtracting(oldApiNames)
+    let deletedApiNames = oldApiNames.subtracting(newApiNames)
+    let persistedApiNames = oldApiNames.intersection(newApiNames)
+    
+    var changes: [String: [ApiChange]] = [:]
+    
+    // Additions
+    
+    for usr in (addedApiNames.map { usr in newApiNameNodeMap[usr]! }.sorted(by: apiNodeIsOrderedBefore)) {
+        let apiType = prettyString(forKind: usr["key.kind"] as! String)
+        let name = prettyName(forApi: usr, apis: newApiNameNodeMap)
+        let root = rootName(forApi: usr, apis: newApiNameNodeMap)
+        changes[root, withDefault: []].append(.addition(apiType: apiType, name: name))
+    }
+    
+    // Deletions
+    
+    for usr in (deletedApiNames.map { usr in oldApiNameNodeMap[usr]! }.sorted(by: apiNodeIsOrderedBefore)) {
+        let apiType = prettyString(forKind: usr["key.kind"] as! String)
+        let name = prettyName(forApi: usr, apis: oldApiNameNodeMap)
+        let root = rootName(forApi: usr, apis: oldApiNameNodeMap)
+        changes[root, withDefault: []].append(.deletion(apiType: apiType, name: name))
+    }
+    
+    // Modifications
+    
+    let ignoredKeys = Set(arrayLiteral: "key.doc.line", "key.parsed_scope.end", "key.parsed_scope.start", "key.doc.column", "key.doc.comment", "key.bodyoffset", "key.nameoffset", "key.doc.full_as_xml", "key.offset", "key.fully_annotated_decl", "key.length", "key.bodylength", "key.namelength", "key.annotated_decl", "key.doc.parameters", "key.elements", "key.related_decls",
+                          "key.filepath", "key.attributes",
+                          "key.parsed_declaration",
+                          "key.docoffset", "key.attributes")
```

Comment:
Here we can configure which type of changes exactly we are interested right now.